src/driver/hermes_cdev.c

```diff
@@ -28,18 +28,199 @@
 #include <linux/module.h>
 #include <linux/pci.h>
 
+#include "hermes.h"
 #include "hermes_mod.h"
+#include "hermes_cdev.h"
+#include "xdma_sgdma.h"
 
 #define HERMES_MINOR_BASE	0
 #define HERMES_MINOR_COUNT	16
 #define HERMES_NAME		"hermes"
 
 static struct class *hermes_class;
 DEFINE_IDA(hermes_ida);
+DEFINE_IDA(hermes_prog_slot_ida);
+DEFINE_IDA(hermes_data_slot_ida);
 static dev_t hermes_devt;
 
+enum hermes_slot_type {
+	HERMES_SLOT_PROG,
+	HERMES_SLOT_DATA,
+};
+
+static void hermes_release_slot(int32_t slot_id,
+				enum hermes_slot_type slot_type)
+{
+	if (slot_id < 0)
+		return;
+
+	if (slot_type == HERMES_SLOT_PROG)
+		ida_simple_remove(&hermes_prog_slot_ida, slot_id);
+	else if (slot_type == HERMES_SLOT_DATA)
+		ida_simple_remove(&hermes_data_slot_ida, slot_id);
+	else
+		pr_err("Invalid slot_type: 0x%x\n", slot_type);
+}
+
+static int32_t hermes_request_slot(struct hermes_env *env,
+				   enum hermes_slot_type slot_type)
+{
+	if (slot_type == HERMES_SLOT_PROG)
+		return ida_simple_get(&hermes_prog_slot_ida, 0,
+				env->hermes->hpdev->cfg.ehpslot, GFP_KERNEL);
+	else if (slot_type == HERMES_SLOT_DATA)
+		return ida_simple_get(&hermes_data_slot_ida, 0,
+				env->hermes->hpdev->cfg.ehdslot, GFP_KERNEL);
+	else
+		return -EINVAL;
+}
+
+static int hermes_open(struct inode *inode, struct file *filp)
+{
+	struct hermes_dev *hermes;
+	struct hermes_env *env;
+
+	hermes = container_of(inode->i_cdev, struct hermes_dev, cdev);
+	env = kzalloc(sizeof(*env), GFP_KERNEL);
+	env->hermes = hermes;
+	env->prog_slot = -1;
+	env->data_slot = -1;
+	filp->private_data = env;
+
+	return 0;
+}
+
+static int hermes_close(struct inode *inode, struct file *filp)
+{
+	struct hermes_env *env = filp->private_data;
+
+	hermes_release_slot(env->prog_slot, HERMES_SLOT_PROG);
+	hermes_release_slot(env->data_slot, HERMES_SLOT_DATA);
+
+	kfree(env);
+
+	return 0;
+}
+
+static ssize_t hermes_read(struct file *filp, char __user *buff, size_t count,
+		loff_t *f_pos)
+{
+	struct hermes_env *env = filp->private_data;
+	struct hermes_pci_dev *hpdev = env->hermes->hpdev;
+	struct xdma_channel *chnl;
+	long res;
+	loff_t pos;
+
+	chnl = xdma_get_c2h(hpdev);
+	if (!chnl) {
+		pr_err("No DMA channel available");
+		return -EIO;
+	}
+
+	if (env->data_slot < 0)
+		return -ENODATA;
+
+	if (count > hpdev->cfg.ehdssze) {
+		pr_warn("Data size greater than data slot size: 0x%lx > 0x%x. Resizing.\n",
+				count, hpdev->cfg.ehdssze);
```

Comment:
Short read requests are normal, you shouldn't warn for them.

Reply:
That's the opposite, reading more than the slot size

Reply:
That's what I mean. If I have a 4KB buffer and read but only get 1KB back (because that's all there is) then it's normal and doesn't require a warning.